Dip/Dip/Definition.swift

```diff
@@ -0,0 +1,66 @@
+//
+//  Definition.swift
+//  Dip
+//
+//  Created by Ilya Puchka on 04.11.15.
+//  Copyright © 2015 AliSoftware. All rights reserved.
+//
+
+import Foundation
+
+///Internal representation of a key used to associate definitons and factories by tag, type and factory.
+struct DefinitionKey : Hashable, Equatable, CustomDebugStringConvertible {
+    var protocolType: Any.Type
+    var factory: Any.Type
```

Comment:
I was confused at first to see that we stored the `factory` closure in the `DefinitionKey`. I now understand that in fact you store the factory _signature_ / Type in there, to store the expected runtime argument types.

So here we probably need to find a different name for this property, to make it less confusing with the other places when we use the term `factory` which expect an actual closure, while in `DefinitionKey` that's the closure type, not the closure itself.
I only understood that when I saw your `let key = DefinitionKey(protocolType: T.self, factory: F.self, associatedTag: tag)` [here](https://github.com/AliSoftware/Dip/pull/8/files#diff-fd56c3307267c4f8aeac89674de93e68R101), seeing that you used `F.self` there.

Simply renaming this to `factoryType` or `factorySignature` or whatnot would make things clearer I think.